### PR TITLE
Allowing storage init to expand P3 to entire disk

### PR DESCRIPTION
**Pull request by rvs**
This change allows creation of live images that are very compact (just the size of the rootfs). The rest of the disk structure is created on the first boot by storage-init.

This is mostly useful when putting EVE on live media that is much bigger.

As the comment says -- this change seems to be pretty safe, since if the P3 partition is not located anywhere -- we may as well try reasonably dangerous actions like resizing the disk.

Finally, this change is next to the last step to allow creating EVE instances as VMs on Google cloud.

P.S. And yest, I do realize that the entire storage-init should probably be re-written in Go -- but this is the task for another time.

### CODE REVIEW

File: pkg/storage-init/storage-init.sh [hunk 1]
```diff
@@ -34,6 +34,57 @@ fi
 
 P3_FS_TYPE="ext3"
 FSCK_FAILED=0
+
+# First lets see if we're missing P3 altogether and try to create it.
+# This is safe, since the worst case scenario we may hose the system,
+# but the system without P3 is a warm brick anyways
+P3=$(/hostfs/sbin/findfs PARTLABEL=P3)
+IMGA=$(/hostfs/sbin/findfs PARTLABEL=IMGA)
```
Comment (deitch) [inline pkg/storage-init/storage-init.sh:42]: This basically says, "if we don't have a `/persist` partition, and we do have IMGA, then calculate where the `/persist` should exist and make it".

If we don't have an IMGA, we probably should be panicking, so maybe let's make this cleaner and even earlier do an if statement that says, `if [ -z "$IMGA" ]; then panic`? That will make this one cleaner and easier to understand.

Also, let's add a comment here as to what the logic is, something like, "if we don't have a persist partition, calculate where it should sit and create it."

File: pkg/storage-init/storage-init.sh [hunk 2]
```diff
@@ -34,6 +34,57 @@ fi
...
+       dd if="$DEV" of="$DEV" bs=1 skip=446 seek=$(( 446 + 16)) count=16 conv=noerror,sync,notrunc
+       # restore 1st MBR entry + first partition entry
+       dd if=/tmp/mbr.bin of="$DEV" bs=1 conv=noerror,sync,notrunc
+
+       # focrce kernel to re-scan partition table
+       partprobe "$DEV"
+   fi
+
+   # lets see if IMGB partition is around, if not - create it
+   IMGB=$(/hostfs/sbin/findfs PARTLABEL=IMGB)
```
Comment (deitch) [inline pkg/storage-init/storage-init.sh:67]: Can we move this up to [here](https://github.com/lf-edge/eve/pull/870/files#diff-043db85bf353990b42dd8c985c767eddR42), so we get all of them at once, easier to read?

File: pkg/storage-init/storage-init.sh [hunk 3]
```diff
@@ -34,6 +34,57 @@ fi
...
+# This is safe, since the worst case scenario we may hose the system,
+# but the system without P3 is a warm brick anyways
+P3=$(/hostfs/sbin/findfs PARTLABEL=P3)
+IMGA=$(/hostfs/sbin/findfs PARTLABEL=IMGA)
+if [ -z "$P3" ] && [ -n "$IMGA" ]; then
+   DEV=$(echo /sys/block/*/"${IMGA#/dev/}")
+   DEV="/dev/$(echo "$DEV" | cut -f4 -d/)"
+
+   # if sgdisk complains we need to repair the GPT
+   if sgdisk -v "$DEV" | grep -q 'Identified.*problems'; then
```
Comment (deitch) [inline pkg/storage-init/storage-init.sh:48]: Dies all of this belong inside the "if persist partition doesn't exist" logic? Isn't this a completely independent piece? Ideally, it should be its own PR, but either way, just make a fn that runs before this called `validateMBR()` or similar?

File: pkg/storage-init/storage-init.sh [hunk 4]
```diff
@@ -34,6 +34,57 @@ fi
...
+
+      IMGA_SIZE=$(sgdisk -i "$IMGA_ID" "$DEV" | awk '/^Partition size:/ { print $3; }')
+      IMGA_GUID=$(sgdisk -i "$IMGA_ID" "$DEV" | awk '/^Partition unique GUID:/ { print $4; }')
+
+      SEC_START=$(sgdisk -f "$DEV")
+      SEC_END=$((SEC_START + IMGA_SIZE))
+      IMGB_ID=$((IMGA_ID + 1))
+
+      sgdisk --new "$IMGB_ID:$SEC_START:$SEC_END" \
+             --typecode="$IMGB_ID:$IMGA_GUID" --change-name="$IMGB_ID:IMGB" "$DEV"
```
Comment (deitch) [inline pkg/storage-init/storage-init.sh:79]: Same thing here. This isn't really part of "if persist partition doesn't exist, create it," but rather its own function that says, "ensure IMGB partition exists". Let's move it up and out to its own thing. Then the "ensure persist partition exists" becomes super clean."

File: pkg/storage-init/storage-init.sh (hunk 1 above)
Comment (rvs) [reply pkg/storage-init/storage-init.sh:42]: Yes. But on the later point -- I don't think we should be  panicking actually -- I think a more sensible strategy is still to allow the box to boot into the shell. Note that whatever booted us is 99.99% a full EVE rootfs image (and yes it may not be marked as IMGA -- but it still is a full image). This means that we can leverage EVE microservices to communicate useful bits and pieces of information back to the cloud.

I fixed the comment.

File: pkg/storage-init/storage-init.sh (hunk 2 above)
Comment (rvs) [reply pkg/storage-init/storage-init.sh:67]: Sure. Done @deitch

File: pkg/storage-init/storage-init.sh (hunk 3 above)
Comment (rvs) [reply pkg/storage-init/storage-init.sh:48]: It does belong there in a sense, that I'm uncomfortable (still!) to touch GPT unless there's something as fundamentally broken as missing P3 or IMGB. So I'll keep it here for now if you don't mind -- but we can consider refactoring it later (especially if we do a more proper implementation of disk resizing along the lines of https://github.com/linuxkit/linuxkit/tree/master/pkg/format)

File: pkg/storage-init/storage-init.sh (hunk 4 above)
Comment (rvs) [reply pkg/storage-init/storage-init.sh:79]: The problem here is the case where P3 does exist but IMGB doesn't. Then it becomes tricky to actually figure out where to put IMGB on disk even. I'd rather keep it here for now and refactor later if you don't mind (simply because I'm not quite sure what a sensible strategy for missing IMGB would be -- bail?)

File: pkg/storage-init/storage-init.sh (hunk 1 above)
Comment (eriknordmark) [reply pkg/storage-init/storage-init.sh:42]: > Yes. But on the later point -- I don't think we should be panicking actually -- I think a more sensible strategy is still to allow the box to boot into the shell. Note that whatever booted us is 99.99% a full EVE rootfs image (and yes it may not be marked as IMGA -- but it still is a full image). This means that we can leverage EVE microservices to communicate useful bits and pieces of information back to the cloud.

@rvs We clearly won't be sending logs to the controller, since we have not /persist/rsyslogd.
Or would we run with /persist in memory, meaning that we can get various subsequent failures which would be hard to diagnose? I think we need to 1) have specific logic to handle the case of no /persist (not watchdog, or run out of memory due to overlayfs) 2) test that the system can operate in this space so that it can at least report to the controller and the controller can send a reboot command.

**Comment by rvs** [comment]
> move some of these functions into their own space, so that the persist partition creation logic is clean and focused

See my comments @deitch -- hopefully for now at least -- they make sense.

> make it clear in the comments that P3 is the persist partition
> add some other comments

Done

> update the build document to indicate how our boot time handles some of the safety checks, 
> and how it is possible to build without persist partition

This one I'm not sure what documentation do you have in mind. URL pleas?

File: pkg/storage-init/storage-init.sh [hunk 5]
```diff
@@ -34,6 +34,57 @@ fi
 
 P3_FS_TYPE="ext3"
 FSCK_FAILED=0
+
+# First lets see if we're missing P3 altogether and try to create it.
+# This is safe, since the worst case scenario we may hose the system,
+# but the system without P3 is a warm brick anyways
```
Comment (eriknordmark) [inline pkg/storage-init/storage-init.sh]: What about the case when IMGB is fine, and eveimage update places a bad version on IMGA which incorrectly decides to reformat the disk, thereby blowing away IMGB, and then the watchdog fires and tries to cause a fallback to the now gone IMGB version?

BTW: Why is this code not symmetric with respect to IMGA/IMGB? IMGA is not special.

**Comment by rvs** [comment]
After chatting with @eriknordmark -- he suggested an even better idea to fence off the code -- lets only run it when both P3 and IMGB are missing. This has a benefit of making it real clear what the condition is under which we are willing to destructively manipulate the partition table AND it makes everything much more symmetrical.
